runtime/src/snapshot_utils.rs

```diff
@@ -649,66 +934,177 @@ pub fn bank_from_snapshot_archive<P>(
+    let mut unpacked_append_vec_map = prepared_full_snapshot_archive.unpacked_append_vec_map;
+    if let Some(ref mut unarchive_preparation_result) = prepared_incremental_snapshot_archive {
+        let incremental_snapshot_unpacked_append_vec_map =
+            std::mem::take(&mut unarchive_preparation_result.unpacked_append_vec_map);
+        unpacked_append_vec_map.extend(incremental_snapshot_unpacked_append_vec_map.into_iter());
+    }
 
+    let mut measure_rebuild = Measure::start("rebuild bank from snapshots");
     let bank = rebuild_bank_from_snapshots(
-        snapshot_version.trim(),
+        &prepared_full_snapshot_archive.unpacked_snapshots_dir_and_version,
+        prepared_incremental_snapshot_archive
+            .as_ref()
+            .map(|unarchive_preparation_result| {
+                &unarchive_preparation_result.unpacked_snapshots_dir_and_version
+            }),
         frozen_account_pubkeys,
-        &unpacked_snapshots_dir,
         account_paths,
         unpacked_append_vec_map,
         genesis_config,
         debug_keys,
         additional_builtins,
-        account_indexes,
+        account_secondary_indexes,
         accounts_db_caching_enabled,
         limit_load_slot_count_from_snapshot,
         shrink_ratio,
         verify_index,
     )?;
-    measure.stop();
+    measure_rebuild.stop();
+    info!("{}", measure_rebuild);
 
-    let mut verify = Measure::start("verify");
+    let mut measure_verify = Measure::start("verify");
     if !bank.verify_snapshot_bank(test_hash_calculation)
         && limit_load_slot_count_from_snapshot.is_none()
     {
         panic!("Snapshot bank for slot {} failed to verify", bank.slot());
     }
-    verify.stop();
+    measure_verify.stop();
+
     let timings = BankFromArchiveTimings {
-        rebuild_bank_from_snapshots_us: measure.as_us(),
-        untar_us: untar.as_us(),
-        verify_snapshot_bank_us: verify.as_us(),
+        rebuild_bank_from_snapshots_us: measure_rebuild.as_us(),
+        full_snapshot_untar_us: prepared_full_snapshot_archive.measure_untar.as_us(),
+        incremental_snapshot_untar_us: prepared_incremental_snapshot_archive
+            .map_or(0, |unarchive_preparation_result| {
+                unarchive_preparation_result.measure_untar.as_us()
+            }),
+        verify_snapshot_bank_us: measure_verify.as_us(),
     };
-
     Ok((bank, timings))
 }
 
+/// Perform the common tasks when unarchiving a snapshot.  Handles creating the temporary
+/// directories, untaring, reading the version file, and then returning those fields plus the
+/// unpacked append vec map.
+fn do_snapshot_unarchive_preparation<P, Q>(
+    snapshots_dir: P,
+    unpacked_snapshots_dir_prefix: &'static str,
+    snapshot_archive_path: Q,
+    measure_name: &'static str,
+    account_paths: &[PathBuf],
+    archive_format: ArchiveFormat,
+    parallel_divisions: usize,
+) -> Result<UnarchivePreparationResult>
```

Comment:
> `Result<UnarchivePreparationResult>`

I think the type sig of this fn is a bit mouthful and I want to avoid 2 nested `Result`s... could you reword this function like this?

```rust
struct UnarchivedSnapshot { // <= renamed from UnarchivePreparationResult
}


fn unarchive_snapshot(...) -> Result<UnarchivedSnapshot> { // renamed from do_snapshot_unarchive_preparation
}
```

better names are welcome. btw, I don't like too reliance of `MetaFooBar`, `FooBarInfo`, `FooBarContext` across our codebase, so I'm eager to remove them. xD

i think enginner's job is to name things properly with good vocabulary for future generation's code readability.. lol

Reply:
I like this, done!